app/uk/gov/hmrc/crdlcacheadminfrontend/auth/AuthController.scala

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2025 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.crdlcacheadminfrontend.auth
+
+import javax.inject.Inject
+import play.api.mvc.MessagesControllerComponents
+import scala.concurrent.Future.successful
+import uk.gov.hmrc.crdlcacheadminfrontend.controllers.routes
+import uk.gov.hmrc.crdlcacheadminfrontend.auth.routes as authRoutes
+import uk.gov.hmrc.internalauth.client.FrontendAuthComponents
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+
+class AuthController @Inject()(auth: FrontendAuthComponents, authActions: AuthActions, mcc: MessagesControllerComponents)
+  extends FrontendController(mcc) {
+    given FrontendAuthComponents = auth
+
+    def signIn() = Action.async { implicit request =>
+        authActions.handle(
+            authRoutes.AuthController.signIn(),
+            r => {
+                successful(Redirect(routes.IndexController.onPageLoad()))
+            }
+        )
+    }
```

Comment:
If we make the `permission` visible then we can use the `authorizedAction` like this:
```suggestion
    def signIn() = auth.authorizedAction(
      continueUrl = authRoutes.AuthController.signIn(),
      predicate   = authActions.permission
    ).async { implicit request =>
      successful(Redirect(routes.IndexController.onPageLoad()))
    }
```
In general `Action`s are a series of functions / filters that run before the body of the request is parsed, so you usually only run one `Action` per request.

The bit inside `async` is our request handling logic after auth and any other cross-cutting stuff has happened.

Reply:
The challenge I am finding is without the outer implicit request, I don't have a subsequent implicitly defined Messages arg for the page call as the request from the authorizedAction is a `AuthenticatedRequest[AnyContent, Unit]` as opposed to a `MessagesRequest[AnyContent]`.
Is there a straightforward way to source this from context that sticks with this pattern?

Reply:
Luckily [there is](https://www.playframework.com/documentation/3.0.x/ScalaI18N#Using-Messages-with-Controllers), although as with all of this implicit stuff unfortunately it's one of those "you just have to know" answers - we have to extend `I18nSupport` in our controller, and this magically converts `RequestHeader` (which all requests extend) into a `Messages`.

Here's [an example](https://github.com/hmrc/object-store-admin-frontend/blob/main/app/uk/gov/hmrc/objectstoreadminfrontend/controllers/ObjectStoreController.scala#L35) from object-store-admin-frontend.

Reply:
Thank you kindly. 🙇